### PR TITLE
introduce region level tombstone sweeper

### PR DESCRIPTION
**Pull request by gesterzhou**
I send this pull request to collect comments and feedback on both design and coding.
It's not complete yet. 

@boglesby @bschuchardt @jhuynh1 @upthewaterspout @agingade @dschneider-pivotal 
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java [hunk 1]
```diff
@@ -119,11 +120,21 @@
...
+  // private Map<Region, NonReplicateTombstoneSweeper> clientRegionSweepers =
+  // Collections.synchronizedMap(new HashMap<Region, NonReplicateTombstoneSweeper>());
+
   public static TombstoneService initialize(InternalCache cache) {
     return new TombstoneService(cache);
   }
 
   private TombstoneService(InternalCache cache) {
+
+    // Keep the existing replicatedTombstoneSweeper to be shared by server side regions
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java]: Why do we keep this generic sweeper if each region has it's own?

File: geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java [hunk 2]
```diff
@@ -119,11 +120,21 @@
   private final ReplicateTombstoneSweeper replicatedTombstoneSweeper;
   private final NonReplicateTombstoneSweeper nonReplicatedTombstoneSweeper;
 
+  // TODO: since we will have region level tombstone sweeper, we might not need
+  // NonReplicateTombstoneSweeper any more
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java]: Is the non replicate for clients?

File: geode-core/src/main/java/org/apache/geode/internal/cache/LocalRegion.java [hunk 3]
```diff
@@ -6383,6 +6393,9 @@ protected void distributeDestroyRegion(RegionEventImpl event, boolean notifyOfRe
    */
   @Override
   public void postCreateRegion() {
+    // add region level tombstone sweeper
+    this.cache.getTombstoneService().addRegionTombstoneSweeper(this);
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/LocalRegion.java]: Does this get created for every bucket region too, or just the PR?
Shouldn't there instead be a way to configure on a per region basis?

File: geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java [hunk 4]
```diff
@@ -163,6 +212,10 @@ public void scheduleTombstone(LocalRegion r, RegionEntry entry, VersionTag destr
 
 
   private TombstoneSweeper getSweeper(LocalRegion r) {
+    TombstoneSweeper sweeper = this.regionTombstoneSweepers.get(r);
+    if (sweeper != null) {
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java]: If every region has it's own sweeper, should it create one if one is somehow missing, or is this a signal that something bad has occurred that this region doesn't have as sweeper?

File: geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java [hunk 5]
```diff
@@ -803,6 +856,11 @@ public void unscheduleTombstones(final LocalRegion r) {
       });
     }
 
+    public void setTOmbstoneTimeout(long expiryTime) {
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java]: Typo, setTombstoneTimeout (the O is capitalized)

File: geode-lucene/src/main/java/org/apache/geode/cache/lucene/internal/LuceneIndexForPartitionedRegion.java [hunk 6]
```diff
@@ -38,18 +38,23 @@
...
 import org.apache.geode.internal.cache.BucketRegion;
 import org.apache.geode.internal.cache.InternalCache;
+import org.apache.geode.internal.cache.LocalRegion;
 import org.apache.geode.internal.cache.PartitionedRegion;
 import org.apache.geode.internal.i18n.LocalizedStrings;
 
 public class LuceneIndexForPartitionedRegion extends LuceneIndexImpl {
   protected Region fileAndChunkRegion;
   protected final FileSystemStats fileSystemStats;
+  protected static long LUCENE_TOMBSTONE_TIMEOUT =
```
Comment (jhuynh1) [inline geode-lucene/src/main/java/org/apache/geode/cache/lucene/internal/LuceneIndexForPartitionedRegion.java]: What does the lucene portion have to do with this commit? should this be in a separate commit?

**Comment by bschuchardt** [comment]
What's the motivation for this PR?

File: geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java [hunk 7]
```diff
@@ -733,11 +786,11 @@ public long getScheduledTombstoneCount() {
     /**
      * the expiration time for tombstones in this sweeper
      */
-    protected final long EXPIRY_TIME;
+    protected long EXPIRY_TIME;
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java]: If possible redesign so these could be "final" again. This would require you to know the tombstone timeout when you create a per region sweeper.

If this is not possible then rename these to not be all caps which is reserved for final constants. Since they were being set in the constructor they are not "constants" anyway (even though they were final) so I recommend you change them to be "expiryTime" and "purgeInterval" anyway.

File: geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java [hunk 8]
```diff
@@ -803,6 +856,11 @@ public void unscheduleTombstones(final LocalRegion r) {
       });
     }
 
+    public void setTOmbstoneTimeout(long expiryTime) {
+      this.EXPIRY_TIME = expiryTime;
+      this.PURGE_INTERVAL = Math.min(DEFUNCT_TOMBSTONE_SCAN_INTERVAL, expiryTime);
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/TombstoneService.java]: If you keep this, allowing these variables to be changed after the constructor, then change the constructor to calll this method instead of duplicating the code that computes PURGE_INTERVAL.

File: geode-lucene/src/integrationTest/java/org/apache/geode/cache/lucene/LuceneIndexCreationPersistenceIntegrationTest.java [hunk 9]
```diff
@@ -77,11 +77,11 @@ public void createCache() {
   }
 
   @Test
-  public void shouldNotUseOverflowForInternalRegionsWhenUserRegionHasOverflow() {
+  public void CanUseOverflowForInternalRegionsWhenUserRegionHasOverflow() {
```
Comment (dschneider-pivotal) [inline geode-lucene/src/integrationTest/java/org/apache/geode/cache/lucene/LuceneIndexCreationPersistenceIntegrationTest.java]: minor nit: method names should start with a lower-case character. Upper case should be reserved for Class/Interface names

File: geode-lucene/src/main/java/org/apache/geode/cache/lucene/internal/LuceneIndexForPartitionedRegion.java [hunk 10]
```diff
@@ -171,15 +176,32 @@ private PartitionResolver getPartitionResolver(PartitionAttributes dataRegionAtt
...
     // Create AttributesFactory based on input RegionShortcut
+    final boolean isOverFlowToDisk = isOverFlowToDisk(regionAttributes);
     RegionAttributes baseAttributes = this.cache.getRegionAttributes(regionShortCut.toString());
     AttributesFactory factory = new AttributesFactory(baseAttributes);
     factory.setPartitionAttributes(partitionAttributesFactory.create());
     if (regionAttributes.getDataPolicy().withPersistence()) {
       factory.setDiskStoreName(regionAttributes.getDiskStoreName());
     }
+    if (isOverFlowToDisk) {
+      factory.setEvictionAttributes(regionAttributes.getEvictionAttributes());
```
Comment (dschneider-pivotal) [inline geode-lucene/src/main/java/org/apache/geode/cache/lucene/internal/LuceneIndexForPartitionedRegion.java]: what does using overflow to disk have to do with per region tombstone sweepers? This should be a seperate checkin

**Comment by metatype** [comment]
Tombstone cleanup has been a sore point for users who want very predictable latency.  How will this change affect latency of ongoing operations like put/get?

**Comment by PurelyApplied** [comment]
This PR hasn't seen any action for over three months.  I'm going to go ahead and close it.  If there is still active work in progress, don't hesitate to reopen it, assuming the above concerns have been addressed.
